File: src/templates/scripts/simple_action_client.py
#!/usr/bin/env python3
import rospy
import actionlib
from custom_msgs.msg import ExampleAction, ExampleGoal

class TemplateActionClient:

    def __init__(self):
        goal = rospy.get_param(param_name='goal')

        self.action_client = actionlib.SimpleActionClient(ns='/example_action', ActionSpec=ExampleAction)

        self.send_goal(goal=goal)
        self.action_client.wait_for_result()

    def send_goal(self, goal):
        self.action_client.wait_for_server()

        goal_msg = ExampleGoal()
        goal_msg.goal = goal

        self.action_client.send_goal(
            goal=goal_msg,
            done_cb=self.done_callback,
            feedback_cb=self.feedback_callback
        )

    def done_callback(self, status, result):
        goal_id = {
            0 : 'PENDING',
            1 : 'ACTIVE',
            2 : 'CANCELED',
            3 : 'SUCCEEDED',
            4 : 'ABORTED',
            5 : 'REJECTED',
            6 : 'CANCELING',
            7 : 'RECALLING',
            8 : 'RECALLED',
            9 : 'LOST'
        }
        
        if status == 3:
            rospy.loginfo(msg=f'Goal succeeded: {result.result}')
        else:
            rospy.logerr(msg=f'Goal failed with status: {goal_id[status]}')

    def feedback_callback(self, feedback):
        rospy.loginfo(msg=f'Feedback: {feedback.feedback}')

if __name__ == '__main__':
    rospy.init_node(name='action_client')
    TemplateActionClient()

File: src/templates/src/simple_action_client.cpp
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <custom_msgs/ExampleAction.h>

class TemplateActionClient
{
    public:

    TemplateActionClient() : action_client("/example_action", true)
    {
        nh.getParam("goal", goal);

        TemplateActionClient::send_goal(goal);        
        action_client.waitForResult();
    }

    void send_goal(int goal)
    {
        action_client.waitForServer();

        custom_msgs::ExampleGoal goal_msg;
        goal_msg.goal = goal;

        action_client.sendGoal
        (
            goal_msg,
            boost::bind(&TemplateActionClient::done_callback, this, _1, _2),
            boost::bind(&TemplateActionClient::active_callback, this),
            boost::bind(&TemplateActionClient::feedback_callback, this, _1)
        );
    }

    void done_callback
    (
        const actionlib::SimpleClientGoalState &state,
        const custom_msgs::ExampleResultConstPtr &result
    )
    {
        if (action_client.getState() == actionlib::SimpleClientGoalState::SUCCEEDED)
        {
            std::string res = (result->result == goal) ? "true" : "false";
            ROS_INFO("Goal succeeded: %s", res.c_str());
        }
        else
        {
            ROS_ERROR("Goal failed with status: %s", action_client.getState().toString().c_str());
        }
    }

    void active_callback() {}

    void feedback_callback(const custom_msgs::ExampleFeedbackConstPtr &feedback)
    {
        ROS_INFO("Feedback: %i", (int)feedback->feedback);
    }

    private:

    ros::NodeHandle nh;
    actionlib::SimpleActionClient<custom_msgs::ExampleAction> action_client;
    int goal;
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "action_client");
    TemplateActionClient action_client;

    return 0;
}